File: 220914/JavaScript/lotto.js
export function countRightNumbers(myNumbers, winNumbers) {
  return myNumbers.reduce((accumulator, number) => (
    winNumbers.includes(number)
      ? accumulator + 1
      : accumulator
  ), 0);
}

export function countZeroNumbers(myNumbers) {
  return myNumbers.reduce((accumulator, number) => (
    number === 0
      ? accumulator + 1
      : accumulator
  ), 0);
}

export function rank(rightNumbersCount, zeroCount) {
  const theLowestRanking = rightNumbersCount === 0
    ? 6
    : 7 - rightNumbersCount;

  const theHighestRanking = rightNumbersCount === 0 && zeroCount === 6
    ? 1
    : theLowestRanking - zeroCount;

  return [theHighestRanking, theLowestRanking];
}

export function solution(myNumbers, winNumbers) {
  const rightNumbersCount = countRightNumbers(myNumbers, winNumbers);

  const zeroCount = countZeroNumbers(myNumbers);

  return rank(rightNumbersCount, zeroCount);
}

File: solutions/Previous/220914/Java/Lotto.java
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class Lotto {
  public int[] solution(int[] myNumbers, int[] winNumbers) {
    List<Integer> myConvertedNumbers = convertToList(myNumbers);

    int matchingNumberCount
        = compareMyNumberAndWinNumber(myConvertedNumbers, winNumbers);

    int zeroCount = countZeroNumber(myConvertedNumbers);

    return rank(matchingNumberCount, zeroCount);
  }

  public List<Integer> convertToList(int[] myNumbers) {
    return Arrays.stream(myNumbers)
        .boxed()
        .collect(Collectors.toList());
  }

  public int compareMyNumberAndWinNumber(
      List<Integer> myConvertedNumbers, int[] winNumbers) {
    int matchingNumberCount = 0;

    for (int winNumber : winNumbers) {
      if (myConvertedNumbers.contains(winNumber)) {
        matchingNumberCount += 1;
      }
    }

    return matchingNumberCount;
  }

  public int countZeroNumber(List<Integer> myConvertedNumbers) {
    return (int) myConvertedNumbers.stream()
        .filter(number -> number == 0)
        .count();
  }

  public int[] rank(int matchingNumberCount, int zeroCount) {
    int theLowestRanking = 7 - matchingNumberCount;
    int theHighestRanking = theLowestRanking - zeroCount;

    if (theLowestRanking == 7) {
      theLowestRanking = 6;
    }

    if (matchingNumberCount == 0 && zeroCount == 0) {
      theHighestRanking = 6;
    }

    return new int[]{theHighestRanking, theLowestRanking};
  }
}
